app/controllers/api/providers_controller.rb

```diff
@@ -61,6 +61,27 @@ def custom_attributes_add_resource(object, type, id, data = nil)
       super(object, type, id, formatted_data)
     end
 
+    def import_vm_resource(type, id = nil, data = {})
+      raise BadRequestError, "Must specify an id for import of VM to a #{type} resource" unless id
+
+      api_action(type, id) do |klass|
+        provider = resource_search(id, type, klass)
+
+        vm_id = parse_id(data['source'], :vms)
+        # check if user can access the VM
+        resource_search(vm_id, :vms, Vm)
+
+        api_log_info("Importing VM to #{provider_ident(provider)}")
+        target_params = {
+          :name       => data['target']['name'],
+          :cluster_id => parse_id(data['target']['cluster'], :clusters),
```

Comment:
`name` and `sparse` are optional and handled on the lower layers. Missing storage and cluster result in `"Couldn't find Storage with 'id'="` and `"Couldn't find EmsCluster without an ID"` respectively.